deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala

```diff
@@ -0,0 +1,275 @@
+  }
+
+  private def findUsedNodesForOutputs(model: ModelProto, newOutputNames: Array[String])
+      : (Array[NodeProto], Array[ValueInfoProto]) = {
+    val graph = model.getGraph
+    val nodes = graph.getNodeList.toArray.map(_.asInstanceOf[NodeProto])
+
+    val allInternalOutputs = nodes.flatMap(node => node.getNodeOutputNames)
+    newOutputNames.foreach(out => {
+      if (!allInternalOutputs.contains(out)) throw new IllegalArgumentException(s"Unknown output: $out")
+    })
+
+    // This is an array which will track which nodes are needed in the new model
+    val nodeUsageStatus = collection.mutable.Map[String, Boolean]()
+    nodes.foreach(node => nodeUsageStatus(node.getName) = false)
+
+    val outputNameToNodeMap = nodes.flatMap(node => node.getNodeOutputNames.map(name => name -> node)).toMap
+
+    // Recursive method to traverse graph backwards, marking nodes as used or not
+    def markAsUsed(node: NodeProto): Unit = {
+      // If the node is already marked, skip it
+      if (!nodeUsageStatus(node.getName)) {
+        nodeUsageStatus(node.getName) = true
+        // This input might be an actual external input variable or initializer, which has no upstream node
+        // Otherwise continue up the graph chain marking other upstream nodes as used.
+        node.getNodeInputNames.foreach(in => {
+          if (outputNameToNodeMap.contains(in)) markAsUsed(outputNameToNodeMap(in))
+        })
+      }
+    }
+
+    // Starting at the outputs we wish to slice at, mark all nodes as needed or not recursively
+    newOutputNames.foreach(out => markAsUsed(outputNameToNodeMap(out)))
+
+    val newNodes = nodes.filter(node => nodeUsageStatus(node.getName))
+    val newOutputs = newOutputNames.map(out => ValueInfoProto.newBuilder().setName(out).build())
+    (newNodes, newOutputs)
+  }
+
+  /*
+   * Construct a GraphProto from a reference graph with a given set of nodes and outputs
+   */
+  private def makeGraph(nodes: Array[NodeProto],
+                        outputs: Array[ValueInfoProto],
+                        source: GraphProto): GraphProto = {
+    val graph = GraphProto.newBuilder(source)
+    graph.clearNode()
+    nodes.foreach(node => graph.addNode(node))
+    graph.clearOutput()
+    outputs.foreach(output => graph.addOutput(output))
+    graph.build()
+  }
+
+  /*
+   * Construct a ModelProto from a given graph and reference model
+   */
+  private def makeModel(graph: GraphProto, source: ModelProto): ModelProto = {
+    val model = ModelProto.newBuilder(source)
+    model.setGraph(graph).build()
+  }
+
+  implicit class AugmentedProtocolStringList(list: ProtocolStringList) {
+    def toStringArray: Array[String] = {
+      list.toArray().map(_.asInstanceOf[String])
+    }
+  }
+
+  implicit class AugmentedNodeProto(node: NodeProto) {
+    def getNodeOutputNames: Array[String] = {
+      node.getOutputList.toStringArray
+    }
+
+    def getNodeInputNames: Array[String] = {
```

Comment:
might want to see if we can make some of these APIs private to our namespace

Reply:
so for the top 2 you want them less private, and the bottom 2 more private?

Reply:
just the bottom 2 and other methods in the class that can be locked down. Erring on side of private is a good idea